File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        
        if (k <= 1) {
            return 0;
        }

        int ans = 0;
        int left = 0;
        int curr = 1;

        for (int right = 0; right < nums.length; right++) {
            curr *= nums[right];
            while (curr >= k) {
                curr /= nums[left];
                left++; // A
            }

            ans += right - left + 1;
        }

        return ans;
        
    }
}

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        curr = 1
        left = 0
        ans = 0
        for right in range(len(nums)):
            curr *= nums[right]
            while curr >= k and left <= right:
                curr //= nums[left]
                left += 1
            ans += right - left + 1
        return ans